Handle stringified numerics

glarizza opened the pull request:
Previously, the type would check the type of value being passed in the DSL
and would use that type when setting the value in the Hocon config file
(i.e. a string value would be a string in the config file, and an array
would be an array value). The problem is that versions of Puppet < 4.0.0
stringify all values passed in the DSL, so it was impossible to pass a true
numeric value via the DSL (since it would be stringified).

This commit updates the type validation to account for this and implements
a munge block to convert the value type to a numeric. A test for this case
has also been added.

cprice404 commented on lib/puppet/type/hocon_setting.rb:
```diff
@@ -44,7 +44,14 @@
+          # Puppet stringifies numerics in versions of Puppet < 4.0.0
+          # Account for this in the type
+          begin
+            numeric_as_string = Integer(value)
```
@glarizza do we need to account for floats?  Would it cause any harm if we were to simply `s/Integer/Float/g` in your PR?

glarizza replied:
Probably wouldn't harm it - I'll make the change.  Also updating the provider with mk_resource_methods instead of explicit getter/setter, so I'll push those changes and a test this morning.

cprice404 commented:
@fpringvaldsen would you mind reviewing this?  And when we hear back from Gary on the float question maybe we can add a commit and another test if he doesn't have time?

glarizza commented:
Please hold off on merging this for two issues:

1.) Using Float() is going to convert integers to floating point numbers with decimal points
2.) 'type' shouldn't be a property, it should be a parameter
3.) This change is turning Floating point numbers into strings

glarizza commented:
Okay, so here's where we're at:

I'm using Float() to handle floating point values, but it will convert all integer values set in the DSL to a float (i.e. the file will contain 13.0 even if 13 is set in the manifest).  The tests don't check for this because 13.0 == 13 in Ruby (though 13.0.eql?(13) is false - so that's the check when/if we differentiate).  We COULD add support for integer vs float in the type property to account for this, or just not care that 13.0 is entered into the file (I'd assume the former is better than the latter).

cprice404 commented:
Could we try to convert to an int first, and if that fails, try float, and then fall back to string?  Then we wouldn't need to add the extra type?

glarizza commented:
Modified the code and added tests.  This should meet our needs.

glarizza commented:
Wait, oops, forgot the munge.  Please hold...

glarizza commented:
Okay, I noticed that the tests didn't catch my omission on the munge block, and that's because there were no tests for the type.  I added some and fixed the munge.  NOW we should be good...

fpringvaldsen commented on spec/unit/puppet/provider/conf_setting/ruby_spec.rb:
```diff
@@ -517,13 +517,41 @@ def validate_file(expected_content,tmpfile = tmpfile)
+      provider = described_class.new(resource)
+      provider.create
+      expect(provider.exists?).to be true
+      #expect(provider.value[0]).to eql(12)
```
I think we should remove this commented line.

glarizza replied:
Done and pushed

On Thu, Apr 30, 2015 at 1:20 PM, Preben Ingvaldsen <notifications@github.com
> wrote:

> In spec/unit/puppet/provider/conf_setting/ruby_spec.rb
> <https://github.com/puppetlabs/puppetlabs-hocon/pull/12#discussion_r29449776>
> :
>
> > +    it "should be able to handle a float value with number type specified" do
> > +      resource = Puppet::Type::Hocon_setting.new(common_params.merge(
> > +                                                     :setting => 'test_key_1.master', :value => 13.37, :type => 'number'))
> > +      provider = described_class.new(resource)
> > +      provider.create
> > +      expect(provider.exists?).to be true
> > +      expect(provider.value[0].eql?(13.37)).to be(true)
> > +    end
> > +
> > +    it "should be able to handle an Integer string value with number type specified" do
> > +      resource = Puppet::Type::Hocon_setting.new(common_params.merge(
> > +                                                     :setting => 'test_key_1.master', :value => '12', :type => 'number'))
> > +      provider = described_class.new(resource)
> > +      provider.create
> > +      expect(provider.exists?).to be true
> > +      #expect(provider.value[0]).to eql(12)
>
> I think we should remove this commented line.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/puppetlabs/puppetlabs-hocon/pull/12/files#r29449776>.
>



-- 
Gary Larizza
Professional Services Engineer
Puppet Labs

http://garylarizza.com

fpringvaldsen commented:
@glarizza Thanks for this! I'm :+1: on this PR, everything looks good to me.

The last thing I think we need to do on this is add some information on the `type` parameter in the README to indicate that the parameter will need to explicitly be set to `'number'` if a numerical value is passed in when a version of Puppet < 4.0.0 is used, but that can go into a subsequent PR.

@cprice404 or @jpinsonault could one of you give this a review?
